logstash-core/lib/logstash/instrument/collector.rb

```diff
@@ -1,79 +1,109 @@
-      @snapsho_rotation_time = ime
+      begin
+        metric = @metric_store.fetch_or_store(namespaces_path, key) do
+          concrete_class(type).new(namespaces_path, key)
+        end
+        metric.execute(*other)
+        changed # we had changes coming in so we can notify the observers
+      rescue MetricStore::ConcurrentMapExpectedError => e
+        logger.error("Collector: Cannot record metric", :exception => e)
+      rescue NameError => e
+        logger.error("Collector: Cannot create concrete class for this metric type",
+                     :type => type,
+                     :namespaces_path => namespaces_path,
+                     :key => key,
+                     :stacktrace => e.backtrace)
+      end
     end
 
-    def self.snapshot_rotation_time
-      @snapshot_rotation_time || SNAPSHOT_ROTATION_TIME
+    # Monitor the `Concurrent::TimerTask` this update is triggered on every successful or not
+    # run of the task, this method will record the execution in the log
+    #
+    # @param [Time] Time of execution
+    # @param [result] Result of the execution
+    # @param [Exception] Exception
+    def update(time_of_execution, result, exception)
+      return true if exception.nil?
+      logger.error("Collector: Something went wrong went sending data to the observers", 
+                   :execution_time => time_of_execution,
+                   :result => result,
+                   :exception => exception)
     end
 
-    private
-    def roll_over?
-      Concurrent.monotonic_time - @last_rotation >= self.class.snapshot_rotation_time
+    # Snapshot the current Metric Store and return it immediately,
+    # This is useful if you want to get acecess to the current metric store without
+    # waiting for a periodic call.
+    # 
+    # @return [LogStash::Instrument::MetricStore]
+    def snapshot_metric
+      @metric_store.dup
     end
 
-    def snapshot
-      # TODO: We currently sent delta to the observers.
-      # Should we keep the whole picture and send the updated state instead?
-      if roll_over? && @snapshot_rotation_mutex.try_lock 
-        # fair rotation of the snapshot done by the winning thread
-        # metric could be written in the previous snapshot.
-        # Since the snapshot isn't written right away
-        # the view of the snapshot should be consistent at the time of
-        # writing, if we don't receive any events for 5 secs we wont send it.
-        # This might be a problem, for time correlation.
-        logger.debug("Collector: Rotating snapshot", :last_rotation => @last_rotation) if logger.debug?
-
-        publish_snapshot
-        rotate_snapshot
-
-        @snapshot_rotation_mutex.unlock
-      end
-
-      @current_snapshot
+    private
+    # Use the string to generate a concrete class for this metrics
+    # 
+    # @param [String] The name of the class
+    # @raise
+    def concrete_class(type)
+      LogStash::Instrument::MetricType.const_get(type.capitalize)
```

Comment:
what happen in the type is something like ```MetricType``` capitalization might not be enough, right?